L80-csharp-load-balancing.md

```diff
@@ -0,0 +1,387 @@
+        public Status Status { get; }
+    }
+    public sealed class CompletionContext {
+        public CompletionContext();
+        public DnsEndPoint? Address { get; set; }
+        public Exception? Error { get; set; }
+    }
+    public sealed class DnsResolver : Resolver {
+        public DnsResolver(Uri address, ILoggerFactory loggerFactory, TimeSpan refreshInterval);
+        public override Task RefreshAsync(CancellationToken cancellationToken);
+        public override void Start(Action<ResolverResult> listener);
+        protected override void Dispose(bool disposing);
+    }
+    public sealed class DnsResolverFactory : ResolverFactory {
+        public DnsResolverFactory(ILoggerFactory loggerFactory, TimeSpan refreshInterval);
+        public override string Name { get; }
+        public override Resolver Create(Uri address, ResolverOptions options);
+    }
+    public abstract class LoadBalancer : IDisposable {
+        protected LoadBalancer();
+        public void Dispose();
+        public abstract void RequestConnection();
+        public abstract void UpdateChannelState(ChannelState state);
+        protected virtual void Dispose(bool disposing);
+    }
+    public abstract class LoadBalancerFactory {
+        protected LoadBalancerFactory();
+        public abstract string Name { get; }
+        public abstract LoadBalancer Create(IChannelControlHelper controller, IDictionary<string, object> options);
+    }
+    public sealed class PickContext {
+        public PickContext();
+        public HttpRequestMessage? Request { get; set; }
+    }
+    public sealed class PickFirstBalancer : LoadBalancer {
+        public PickFirstBalancer(IChannelControlHelper controller, ILoggerFactory loggerFactory);
+        public override void RequestConnection();
+        public override void UpdateChannelState(ChannelState state);
+        protected override void Dispose(bool disposing);
+    }
+    public sealed class PickFirstBalancerFactory : LoadBalancerFactory {
+        public PickFirstBalancerFactory(ILoggerFactory loggerFactory);
+        public override string Name { get; }
+        public override LoadBalancer Create(IChannelControlHelper controller, IDictionary<string, object> options);
+    }
+    public sealed class PickResult {
+        public Status Status { get; }
+        public Subchannel? Subchannel { get; }
+        public PickResultType Type { get; }
+        public static PickResult ForDrop(Status status);
+        public static PickResult ForFailure(Status status);
+        public static PickResult ForQueue();
+        public static PickResult ForSubchannel(Subchannel subchannel, Action<CompletionContext>? onComplete = null);
+        public void Complete(CompletionContext context);
+    }
+    public abstract class Resolver : IDisposable {
+        protected Resolver();
+        public void Dispose();
+        public abstract Task RefreshAsync(CancellationToken cancellationToken);
```

Comment:
This method returns a `Task`. An implementor doesn't need to do anything with the cancellation token argument. It is free to ignore it.

A no-op implementation of refresh async is:

```cs
public override Task RefreshAsync(CancellationToken cancellationToken)
{
    return Task.CompletedTask;
}
```

Reply:
> But this API now requires that the channel also track whether or not there is currently a refresh operation in flight, so that when it needs to shut down the resolver, it will need to both use this cancellation token and call Dispose(). What's the benefit of requiring the channel to do that, as opposed to just calling Dispose() and having the resolver automatically cancel whatever operations it may currently have in flight?

The channel doesn't need to keep track of an in-progress refresh. A channel simply has a [`CancellationTokenSource`](https://docs.microsoft.com/en-us/dotnet/api/system.threading.cancellationtokensource?view=net-5.0) and when it calls RefreshAsync is just passes `CancellationTokenSource.Token` to the method. When the channel shuts down, it just calls `CancellationTokenSource.Cancel()`. A simple fire-and-forget.

Perhaps this pattern is more difficult and restrictive in C++, but it is really simple in .NET. Typically every async method will take a CancellationToken argument. It's best practice.

Pseudo channel implementation:

```cs
public class Channel
{
    private CancellatonTokenSource _cts = new CancellationTokenSource();
    private Resolver _resolver = CreateResolver();

    public async Task ConnectAsync()
    {
        await _resolver.RefreshAsync(_cts.Token);
    }

    public async Task ShutdownAsync()
    {
        _cts.Cancel();
        _resolver.Dispose();
    }
}
```

You're right that we are signaling that a resolver should stop when dispose is called on it, and CancellationToken on RefreshAsync isn't 100% necessary, but that forces everyone who creates a resolver and has async work in RefreshAsync to create and manage their own CTS. Not a big deal - CTSs are easy to work with - but the cost to the channel managing it for the resolver is basically zero so it makes life slightly easier for resolver authors.

Reply:
> The channel doesn't need to keep track of an in-progress refresh. A channel simply has a [`CancellationTokenSource`](https://docs.microsoft.com/en-us/dotnet/api/system.threading.cancellationtokensource?view=net-5.0) and when it calls RefreshAsync is just passes `CancellationTokenSource.Token` to the method. 

Ah, okay, I had this backwards.  For some reason, I was thinking that this was saying that the resolver returned the cancellation token to the channel, not that the channel passed it to the resolver.

> When the channel shuts down, it just calls `CancellationTokenSource.Cancel()`. A simple fire-and-forget.

That's actually the exact opposite of fire-and-forget.  That's fire-and-you-need-to-remember.

> You're right that we are signaling that a resolver should stop when dispose is called on it, and CancellationToken on RefreshAsync isn't 100% necessary, but that forces everyone who creates a resolver and has async work in RefreshAsync to create and manage their own CTS.

This still does not make any sense.  No caller ever "has async work in RefreshAsync", because *there is no work associated with this method*.  This method is supposed to be fire-and-forget; it's a signal to the resolver that the channel wants fresh data if the resolver can get it, but the resolver has to decide whether to do anything with that signal, and the caller is not waiting on any response.  The caller sends the signal and then goes on with its business, not knowing or caring whether it will ever get another response.

I do understand what you're saying about C# async API semantics.  But the point I'm trying to make here is that this is *not* actually an async API, so I don't think that logic actually applies here.

Reply:
> This method is supposed to be fire-and-forget; it's a signal to the resolver that the channel wants fresh data if the resolver can get it, but the resolver has to decide whether to do anything with that signal, and the caller is not waiting on any response. The caller sends the signal and then goes on with its business, not knowing or caring whether it will ever get another response.

`Resolver.RefreshAsync` is invoked by the channel without awaiting the task. The `Task` returned from the refresh method is then stored by the channel, and the load balancer continues on its way. So it is fire-and-forget.

There are a bunch of reasons for refresh to return a task:

1. Having the task is useful because it allows the channel to know whether the previous refresh is still in-progress. That avoids the need for each resolver to have code to track whether it is running and skip doing anything if refresh is called while still running. The channel does this for you (if a resolver implementation doesn't want this behavior and *wants* to be called while still running then it can return a completed task).
For example: https://github.com/grpc/grpc-java/blob/881747a63df79a7baae252de86447c0997c76f69/core/src/main/java/io/grpc/internal/DnsNameResolver.java#L385-L388

2. Having an extension point on resolver that returns a `Task` is also important because it provides a place for implementors to override and add their async logic to do service discovery. Example of an implementation that loads from filesystem here: https://docs.microsoft.com/en-us/aspnet/core/grpc/loadbalancing?view=aspnetcore-6.0#create-a-custom-resolver
If the extension point doesn't return `Task` then implementors will be forced to do `async void`, which is considered bad practice in async code. See https://stackoverflow.com/a/45449457/11829 for reasons.

3. A returned task provides a way to log errors thrown out of resolver user code.

Reply:
I think what you raised in the last post helped me rethink this. Changes that may addresses your concerns:

```diff
    public abstract class Resolver : IDisposable {
        protected Resolver();
        public void Dispose();
        protected virtual void Dispose(bool disposing);
-       public abstract Task RefreshAsync(CancellationToken cancellationToken);
-       public abstract void Start(Action<ResolverResult> listener);
+       public void Refresh();
+       public void Start(Action<ResolverResult> listener);
+       protected abstract Task ResolveAsync(CancellationToken cancellationToken);
+       protected virtual void OnStart();
+       protected Action<ResolverResult> Listener { get; }
    }
```

I think a big part of the problem here is the previous design [followed what grpc-java does](https://github.com/grpc/grpc-java/blob/b746bab97b6910aa55bd8f0a40d8f5b8b51501cb/api/src/main/java/io/grpc/NameResolver.java#L115-L127). `NameResolver.refresh` is the extension point in Java. If we do the same thing in .NET, but make refresh async, then that async machinery (task and cancellation token) leaks out of the resolver abstraction.

Changes:
* `Refresh` is no longer async. It no longer returns a task or takes a cancellation token. It also isn't virtual. Its job is to invoke a `ResolveAsync`.
* Add a new virtual `ResolveAsync` method. That is where the implementation lives. It is passed a `CancellationToken` when it is called that is internal to the resolver. The token lifetime is tied to the resolver. It is canceled when the resolver is disposed of. Errors out of this method are caught and logged.
* Refactored start so that the listener is saved for you. Start logic, i.e. starting a timer, can optionally be added by overriding `OnStart`.

Thoughts?

Reply:
This definitely makes the public API much cleaner!

How will this work for a watch-based resolver like xDS, where there are no discrete resolution attempts to perform?  As a very simple example, consider the case of a watch-based resolver that gets its data using the following streaming RPC API:

```
// Client request message, sent once at the start of the stream to tell the
// server what name the client is subscribing to updates for.
message ResolveNameRequest {
  string name_to_resolve = 1;
}

// Server response message.  The server will send one response immediately
// upon receiving the client's request.  It will then send subsequent responses
// whenever the set of addresses changes.
message ResolveNameResponse {
  message Address {
    string ip_address = 1;
    int32 port = 2;
  }
  repeated Address addresses = 1;
}

service NameService {
  rpc ResolveName(ResolveNameRequest) returns (stream ResolveNameResponse);
}
```

A resolver that uses this interface will want its `Start()` method to establish the RPC stream to the server, which I think can be done via `OnStart()`.  It will then wait for responses on that stream, and every time it receives such a response, it will report a new result to the channel.  If the stream ends due to a network error or the server closing it, the resolver will reestablish it.  The client will not close the stream until it is disposed.  The `Refresh()` method will be a no-op.

Note that a watch-based resolver like this one does not have any concept of an individual resolution attempt, because it does not make individual requests.  The stream needs to be started when the resolver starts, but after that there is nothing else for it to initiate: there is no periodic timer to start subsequent resolution attempts, there is no backoff and retry for an individual failed resolution attempt, and there is no resolution attempt triggered by a refresh from the channel, so there is no need to ensure that there is only one resolution request going on at any particular time.

It seems like this interface requires the resolver to implement `ResolveAsync()` to provide most of the functionality, but it's not clear to me what this kind of watch-based resolver implementation would do in that method, or whether that would be sufficient to implement its functionality.

Another example to consider here is a resolver for the `ipv4:` URI scheme, which can be useful for specifying a hard-coded address in tests (e.g., `ipv4:127.0.0.1:80`).  In that case, there's no async I/O at all; when the `Start()` method is called, it will immediately return its hard-coded address, and it will never return another result.  It just sits there doing nothing until it is disposed.

There are also cases in tests where it's useful to build a fake resolver for dependency injection, where the test can tell the resolver to return a specific set of addresses at a particular time.  Again there is no async work for the resolver to do, and `Refresh()` is a no-op.

Can this interface handle all of these use-cases?

Reply:
In all the situations you mentioned `ResolveAsync()` would do nothing. It would be overridden to return a completed task and do nothing else.

Example of this: https://github.com/grpc/grpc-dotnet/blob/5d447fdc8edc8bdafb1a5ee4ea763a7c5cb1b157/src/Grpc.Net.Client/Balancer/StaticResolver.cs#L30-L62

It could be made virtual and have that as a default implementation. I'll think about that.

Reply:
If we know that not all subclasses are going to need this method, why make it part of the interface at all?  It seems like this is hard-coding assumptions about the implementation into the interface, which unnecessarily complicates things for resolvers for which that assumption is untrue.

As an alternative, if your goal here is to provide common useful functionality so that individual polling-based resolvers don't need to reinvent the wheel, how about splitting this into two classes?  You can keep the `Resolver` interface very generic, with only `Start()` and `Refresh()` methods, both of which are virtual, and `Refresh()` would be a no-op if not overridden.  Then you can have a separate `PollingResolver` class that inherits from `Resolver` and provides common functionality for polling-based resolvers.  `PollingResolver` would provide the default `Start()` and `Refresh()` implementations you have here, and it would have the protected `ResolveAsync()` method.  That way, polling-based resolvers like the DNS resolver can inherit from `PollingResolver`, while static or watch-based resolvers can inherit directly from `Resolver` and won't have to implement methods they don't need.

Reply:
The issue isn't polling vs non-polling. For example, if DnsResolver didn't have polling, it would still need ResolveAsync to react to Refresh request calls.

I think the best option here is to make `ResolveAsync()` virtual instead of abstract. Implementers who don't need it don't need to worry about it.

Reply:
I experimented with your idea and split `AsyncResolver` out of `Resolver`.

Resolver is now basically identical to Java's:
```csharp
    public abstract class Resolver : IDisposable {
        protected Resolver();
        public void Dispose();
        public abstract void Refresh();
        public abstract void Start(Action<ResolverResult> listener);
        protected virtual void Dispose(bool disposing);
    }
```

Reply:
This looks a lot better!

One clarification, though:

> The issue isn't polling vs non-polling. For example, if DnsResolver didn't have polling, it would still need ResolveAsync to react to Refresh request calls.

I think you may not have understood what I meant by "polling" here.

There are two main types of resolvers, polling-based and watch-based.  A polling-based resolver is one like DNS, where the client decides when to make an individual request and gets back a one-time response, and after that it does not get any updates until it decides to make another request.  A watch-based resolver, on the other hand, is one like the example I described earlier in this thread, where the client subscribes to a name and will be given updates whenever that name's address(es) change; it does not need to decide when to make individual requests, which means that there is no need for refreshes or for waiting for the results of any one async request.

Given that, a DNS resolver is *always* a polling-based resolver; that's inherent in how DNS as a mechanism works.  The `ResolveAsync()` method is relevant only for polling-based resolvers; a watch-based resolver will have no use for it, because it is not making discrete requests.

So I think the issue here is actually about polling-based vs. watch-based resolvers.  The purpose of the `AsyncResolver` class seems to be to provide a convenient framework for implementing polling-based resolvers -- it provides all of the infrastructure for deciding when to trigger an individual request, so the subclass only needs to implement is a single method for actually triggering that request.  Given that, I would suggest renaming it to `PollingResolver`.  I think `AsyncResolver` seems a little misleading, since watch-based resolvers also return results in an async way; they just obtain those results in a different way.

The `Resolver` interface looks much better now; it provides the right interface for use from the channel, and it should work for both polling-based and watch-based mechanisms.  One thing you may want to consider is making the `Refresh()` method be a no-op by default, but allowing subclasses to implement it if they choose.  That way, watch-based resolver implementations don't have to explicitly define that method as a no-op.  But alternatively, if you think that there may be other polling-based resolvers that will inherit directly from `Resolver` instead of using `AsyncResolver`, then you might want to keep it abstract, so that you force polling-based resolver impls to decide how they want to handle it instead of accidentally failing to notice that they need to support it.

Reply:
Ok, I see what you mean by polling.

I'd prefer to leave the resolvers unchanged. Reasoning:
* You're right that `ResolveAsync()` is only relevant to polling-based resolvers, but it is also only useful if the polling has async. Non-async polling that returns a completed task from `ResolverAsync()` is just `Refresh()` but extra work in the background. Non-async polling implementations can inherit from the base `Resolver`.
* `Resolver.Refresh()` could be given a default no-op implementation. Not all resolvers use it. The same could also be said for resolvers that might not need `Resolver.Start(...)`. I'd rather keep both abstract and make the implementor think about them rather than making one or both default no-op implementations.

Reply:
I'll leave this up to you, but I disagree with your reasoning:

> * You're right that `ResolveAsync()` is only relevant to polling-based resolvers, but it is also only useful if the polling has async. Non-async polling that returns a completed task from `ResolverAsync()` is just `Refresh()` but extra work in the background. Non-async polling implementations can inherit from the base `Resolver`.

I don't believe that the concept of "non-async polling implementation" actually exists. :)  I think that in any case where you have an external source of data that you need to poll periodically, that polling will by definition need to be done asynchronously.  The only case I can think of where the data is available synchronously is something like the `ipv4` resolver I mentioned earlier, where no polling is necessary in the first place.  I would be interested in hearing of a counter-example, if you have one in mind.

> * `Resolver.Refresh()` could be given a default no-op implementation. Not all resolvers use it. The same could also be said for resolvers that might not need `Resolver.Start(...)`. I'd rather keep both abstract and make the implementor think about them rather than making one or both default no-op implementations.

I don't think it's actually possible to have a resolver impl where `Start()` is a no-op; that method will always need to be implemented.  If it's not, then the resolver will never return any results, and the channel will just queue all RPCs forever (or until their deadline, if they have one).

Reply:
> I don't think it's actually possible to have a resolver impl where Start() is a no-op; that method will always need to be implemented. If it's not, then the resolver will never return any results, and the channel will just queue all RPCs forever (or until their deadline, if they have one).

Of course, you're right. I'll change it.

Reply:
> The only case I can think of where the data is available synchronously is something like the ipv4 resolver I mentioned earlier, where no polling is necessary in the first place. I would be interested in hearing of a counter-example, if you have one in mind.

There are some synchronous blocking APIs in .NET that could be used without async (in .NET async refers to APIs that return a task, aka promise/future). However, blocking the thread in Refresh would be a buggy implementation.

I thought about it more and I'll go ahead with renaming `AsyncResolver` to `PollingResolver`.

Reply:
This looks great!